### PR TITLE
Segwit Support

### PR DESCRIPTION
**Pull request by hewigovens**
Script\Type | P2PK | P2PKH | P2WPKH | P2SH P2WPKH | P2WSH | P2SH P2WSH
--------|--------|--------|--------|--------|--------|--------
redeem | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ 
lock | N/A | ✅ | ✅ | ✅ | ✅ | N/A

### CODE REVIEW

File: Tests/Bitcoin/BitcoinSignerTests.swift [hunk 1]
```diff
@@ -41,17 +42,20 @@ class BitcoinSignerTests: XCTestCase {
...
+    func testRedeemScriptWrong() {
+        let data = Data(hexString: "042de45bea3dada528eee8a1e04142d3e04fad66119d971b6019b0e3c02266b79142158aa83469db1332a880a2d5f8ce0b3bba542b3e32df0740ccbfb01c275e42")!
+        let redeemHash = Crypto.sha256ripemd160(data)
+        XCTAssertEqual(redeemHash.hexString, "cf5007e19af3641199f21f3fa54dff2fa2627471")
+    }
 
+    func testRedeemScript() {
         let publicKey = PublicKey(data: Data(hexString: "042de45bea3dada528eee8a1e04142d3e04fad66119d971b6019b0e3c02266b79142158aa83469db1332a880a2d5f8ce0b3bba542b3e32df0740ccbfb01c275e42")!)!
-        XCTAssertEqual(publicKey.bitcoinAddress(prefix: 0x05).description, "3LbBftXPhBmByAqgpZqx61ttiFfxjde2z7")
+        let address = publicKey.bitcoinAddress(prefix: 0x05)
```
Comment (vikmeup) [inline Tests/Bitcoin/BitcoinSignerTests.swift]: should we do `0x05` => `Bitcoin().payToScriptHashAddressPrefix`?

File: Tests/Bitcoin/BitcoinSignerTests.swift [hunk 2]
```diff
@@ -122,6 +122,44 @@ class BitcoinSignerTests: XCTestCase {
         //XCTAssertEqual(serialized.hexString, "01000000000102fe3dc9208094f3ffd12645477b3dc56f60ec4fa8e6f5d67c565d1c6b9216b36e000000004847304402200af4e47c9b9629dbecc21f73af989bdaa911f7e6f6c2e9394588a3aa68f81e9902204f3fcf6ade7e5abb1295b6774c8e0abd94ae62217367096bc02ee5e435b67da201ffffffff0815cf020f013ed6cf91d29f4202e8a58726b1ac6c79da47c23d1bee0a6925f80000000000ffffffff0100f2052a010000001976a914a30741f8145e5acadf23f751864167f32e0963f788ac000347304402200de66acf4527789bfda55fc5459e214fa6083f936b430a762c629656216805ac0220396f550692cd347171cbc1ef1f51e15282e837bb2b30860dc77c8f78bc8501e503473044022027dc95ad6b740fe5129e7e62a75dd00f291a2aeb1200b84b09d9e3789406b6c002201a9ecd315dd6a0e632ab20bbb98948bc0c6fb204f2c286963bb48517a7058e27034721026dccc749adc2a9d0d89497ac511f760f45c47dc5ed9cf352a58ac706453880aeadab210255a9626aebf5e29c0e6538428ba0d1dcf6ca98ffdf086aa8ced5e0d0215ea465ac00000000")
     }
 
+    func testSignP2SH_P2WPKH() {
+        let redeemScript = BitcoinScript(data: Data(hexString: "001479091972186c449eb1ded22b78e40d009bdf0089")!)
+        let scripthash = Crypto.sha256ripemd160(redeemScript.data)
```
Comment (alejandro-isaza) [inline Tests/Bitcoin/BitcoinSignerTests.swift]: This should be `79091972186c449eb1ded22b78e40d009bdf0089`

**Comment by hewigovens** [comment]
compiling on 32bit devices failed: 

```bash
Testing failed:
	Cannot convert value of type '(UnsafeRawBufferPointer) -> Data' to expected argument type '(UnsafeRawBufferPointer) -> _'
	Cannot convert value of type '(UnsafeRawBufferPointer) -> Data' to expected argument type '(UnsafeRawBufferPointer) -> _'
** TEST FAILED **
The following build commands failed:
	CompileSwift normal i386 /Users/travis/build/TrustWallet/trust-core/Sources/Bitcoin/BitcoinTransactionSigner.swift
	CompileSwiftSources normal i386 com.apple.xcode.tools.swift.compiler
(2 failures)
The command "xcodebuild clean test -workspace TrustCore.xcworkspace -scheme "TrustCore" -destination 'platform=iOS Simulator,name=iPhone 5'" exited with 65.
```

File: Sources/Bitcoin/BitcoinScript.swift [hunk 3]
```diff
@@ -223,7 +223,7 @@ public final class BitcoinScript: BinaryEncoding, CustomDebugStringConvertible {
         var data = Data(capacity: 5 + pubKeyHash.count)
         data.append(contentsOf: [OpCode.OP_DUP, OpCode.OP_HASH160])
         data.append(UInt8(pubKeyHash.count))
-        data.append(pubKeyHash)
+        data.append([UInt8](pubKeyHash), count: pubKeyHash.count)
```
Comment (alejandro-isaza) [inline Sources/Bitcoin/BitcoinScript.swift]: I don't understand why this fixes anything 🤔

Comment (vikmeup) [reply Sources/Bitcoin/BitcoinScript.swift]: Swift mystery. 

Remember we had similar bug? Data wasn’t converted properly

Comment (alejandro-isaza) [reply Sources/Bitcoin/BitcoinScript.swift]: Right

**Pull request merged**